### PR TITLE
IRC: append build url to failed/successful build

### PR DESCRIPTION
**Pull request by makefu**
## Contributor Checklist:

* [x] I have updated the unit tests
* [x] I have created a file in the master/buildbot/newsfragment directory (and read the README.txt in that directory)
* [NA] I have updated the appropriate documentation

the retiolum workflow requires buildbot to yield the URL of the last build to track build failures

### CODE REVIEW

**Comment by tardyp** [comment]
please update the unit tests

**Comment by sa2ajj** [comment]
LGTM, otherwise

**Comment by makefu** [comment]
@sa2ajj @tardyp i will try my best to build a new test for the url if set but the whole test setup is quite complex - learning curve is similar to `vim` :P

**Comment by tardyp** [comment]
@makefu the test setup is quite simple:

follow the following tutorial http://trac.buildbot.net/wiki/RunningBuildbotWithVirtualEnv

use the Install from source part.

then run

    trial buildbot.test.unit.test_reporters_irc

**Comment by makefu** [comment]
@tardyp ok done, please check if the changes are okay. especially regarding the extension of fakemaster. I was a bit confused with this as i expected that the original implementation of master.status is in used - not a mocked one ;)

File: master/buildbot/reporters/words.py [hunk 1]
```diff
@@ -550,25 +550,22 @@ def buildFinished(self, build):
...
         # FIXME: where do we get the list of changes for a build ?
         # if self.bot.showBlameList and buildResult != SUCCESS and len(build.changes) != 0:
         #    r += '  blamelist: ' + ', '.join(list(set([c.who for c in build.changes])))
-
-        # FIXME: where do we get the base_url? Then do we use the build Link to
-        # make the URL?
-        buildurl = None  # self.bot.status.getBuildbotURL() + build
-        if buildurl:
-            self.send("Build details are at %s" % buildurl)
+        buildurl = self.master.status.getURLForBuild(builder['builderid'], buildNumber)
```
Comment (tardyp) [inline master/buildbot/reporters/words.py]: Sorry I did not see that.
Please dont use master.status. please use reporter.util package. You can see lots of example in the other reporters (like http.py)

File: master/buildbot/test/unit/test_reporters_words.py [hunk 2]
```diff
@@ -518,7 +518,19 @@ def test_command_watch_builder0_get_notifications(self):
...
         self.assertEqual(len(self.sent), 1)
         self.assertIn(
-            'Hey! build builder1 #6 is complete: Success []', self.sent)
+            'Build builder1 #6 is complete: Success []', self.sent)
+
+    @defer.inlineCallbacks
+    def test_command_watch_builder0_get_notifications_with_url(self):
+        # (continue from the prior test)
+        self.bot.tags = None
+        self.master.status.setBuildbotURL("http://example.com")
```
Comment (tardyp) [inline master/buildbot/test/unit/test_reporters_words.py]: you can set it in self.master.config.buildbotURL

File: master/buildbot/test/fake/fakemaster.py [hunk 3]
```diff
@@ -66,6 +66,7 @@ class FakeStatus(service.BuildbotService):
 
     name = "status"
     lastBuilderStatus = None
+    buildbotURL = None
```
Comment (tardyp) [inline master/buildbot/test/fake/fakemaster.py]: this is ancient API. Please dont use it

File: master/buildbot/test/unit/test_reporters_words.py (hunk 2 above)
Comment (makefu) [inline master/buildbot/test/unit/test_reporters_words.py]: this is not used by the mocked status .

File: master/buildbot/reporters/words.py (hunk 1 above)
Comment (makefu) [inline master/buildbot/reporters/words.py]: well this is a weird choice of API change ... but ok i will do as you wish.
maybe add an Obsolete warning to old APIs?

**Comment by makefu** [comment]
@tardyp i switched to the new `reporter.utils` API but right now setting buildbotURL to None results in an exception in `reporters.utils.getURLForBuild`, see `test_command_watch_builder0_get_notifications_without_url`.
Is an unset URL currently even supported or is this use case not applicable anymore? If so, i'd rather remove the new test case.

File: master/buildbot/test/unit/test_reporters_words.py [hunk 4]
```diff
@@ -518,7 +518,8 @@ def test_command_watch_builder0_get_notifications(self):
         yield self.sendBuildFinishedMessage(16)
         self.assertEqual(len(self.sent), 1)
         self.assertIn(
-            'Hey! build builder1 #6 is complete: Success []', self.sent)
+                "Build builder1 #6 is complete: Success [] - "
```
Comment (tardyp) [inline master/buildbot/test/unit/test_reporters_words.py]: I am not sure exactly what this [] is for

Comment (makefu) [inline master/buildbot/test/unit/test_reporters_words.py]: This comes from the master. in the IRC message it contains either red or green state_string (e.g. finished,abort,etc) if colorization is enabled. If not, the status code is plain text. see https://github.com/buildbot/buildbot/blob/master/master/buildbot/reporters/words.py#L559

In the test case the status seems to return an uncolored empty string.

**Pull request merged**